core/shared/platform/common/posix/posix_thread.c

```diff
@@ -47,6 +54,166 @@ os_thread_wrapper(void *arg)
     return NULL;
 }
 
+static void
+stack_context_sig_handler(int sig)
+{
+    assert(sig == STACK_CTX_SIG);
+    korp_tid self = pthread_self();
+
+    /* Get latest stack context (first in the list) for current thread */
+    bh_list *context_list = bh_hash_map_find(stack_context_handler.contexts,
+                                             (void *)(uintptr_t)self);
+    assert(context_list);
+    sigjmp_buf *context = bh_list_first_elem(context_list);
+    assert(context);
+
+    siglongjmp(*context, 1);
+}
+
+sigjmp_buf *
+os_create_stack_context(korp_tid handle)
+{
+    return BH_MALLOC(sizeof(sigjmp_buf));
+}
+
+static void
+free_stack_context_list(void *stack_context_list)
+{
+    bh_list *context_list = (bh_list *)stack_context_list;
+
+    sigjmp_buf *context = bh_list_first_elem(context_list);
+    while (context) {
+        sigjmp_buf *next = bh_list_elem_next(context);
+        BH_FREE(context);
+        context = next;
+    };
+
+    BH_FREE(context_list);
+}
+
+static inline void
+stack_ctx_block_sig_and_lock()
+{
+    sigemptyset(&newmask);
+    sigaddset(&newmask, STACK_CTX_SIG);
+    pthread_sigmask(SIG_BLOCK, &newmask, &oldmask);
+
+    os_mutex_lock(&stack_context_handler.lock);
+}
+
+static inline void
+stack_ctx_unlock_and_unblock_sig()
+{
+    os_mutex_unlock(&stack_context_handler.lock);
+    pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
+}
+
+bool
+os_stack_contexts_init()
+{
+    struct sigaction act;
+    memset(&act, 0, sizeof(act));
+    act.sa_handler = stack_context_sig_handler;
+    sigfillset(&act.sa_mask);
+    if (sigaction(STACK_CTX_SIG, &act, NULL) < 0) {
+        LOG_ERROR("failed to set signal handler");
+        return false;
+    }
+
+    if (os_mutex_init(&stack_context_handler.lock) != 0) {
+        LOG_ERROR("failed to init context mutex");
+        return false;
+    }
+
+    stack_context_handler.contexts = bh_hash_map_create(
+        32, false, (HashFunc)thread_handle_hash,
+        (KeyEqualFunc)thread_handle_equal, NULL, free_stack_context_list);
+    if (stack_context_handler.contexts == NULL) {
+        LOG_ERROR("failed to init stack contexts");
+        return false;
+    }
+
+    return true;
+}
+
+void
+os_stack_contexts_destroy()
+{
+    stack_ctx_block_sig_and_lock();
+    if (!bh_hash_map_destroy(stack_context_handler.contexts))
+        LOG_ERROR("failed to destroy stack context map");
+    stack_context_handler.contexts = NULL;
+    stack_ctx_unlock_and_unblock_sig();
+
+    if (os_mutex_destroy(&stack_context_handler.lock) != BHT_OK)
+        LOG_ERROR("failed to destroy stack contexts");
+}
+
+void
+os_remove_stack_context(korp_tid handle, sigjmp_buf *context)
+{
+    korp_tid self = pthread_self();
+
+    stack_ctx_block_sig_and_lock();
```

Comment:
why don't you make it thread-local?

Reply:
Good point, otherwise it would be shared between threads (and would then require access protection).

Actually, I wanted to avoid globals and return the `oldmask` from `stack_ctx_block_sig_and_lock()`, but returning anything from that function was corrupting the stack when used inside `os_is_returning_from_signal()`, I'm still not sure I understand why